dll/ntdll/ldr/ldrutils.c

```diff
@@ -336,13 +236,105 @@ LdrpUpdateLoadCount3(IN PLDR_DATA_TABLE_ENTRY LdrEntry,
                             if (ShowSnaps)
                             {
-                                DPRINT1("LDR: Flags %lu  %wZ (%lx)\n", Flags, ImportNameUnic, Entry->LoadCount);
+                                DPRINT1("LDR: Flags %lu  %wZ (%lx)\n", Flags, RedirectedImportName, Entry->LoadCount);
                             }
                         }
 
                         /* Recurse into this entry */
                         LdrpUpdateLoadCount3(Entry, Flags, UpdateString);
                     }
+                    else if (RedirectedDll)
+                    {
+                        DPRINT1("LDR: LdrpCheckForLoadedDll failed for redirected dll %wZ\n", RedirectedImportName);
+                    }
+                }
+                else
+                {
+                    /* Unrecoverable SxS failure */
+                    DPRINT1("LDR: RtlDosApplyFileIsolationRedirection_Ustr failed for dll %wZ\n", ImportNameUnic);
+                }
+
+            }
+
+            /* Go through forwarders */
+            NewImportForwarder = (PIMAGE_BOUND_FORWARDER_REF)(BoundEntry + 1);
+            for (i = 0; i < BoundEntry->NumberOfModuleForwarderRefs; i++)
+            {
+                ImportName = (LPSTR)FirstEntry + NewImportForwarder->OffsetModuleName;
+
+                RtlInitAnsiString(&ImportNameAnsi, ImportName);
+                Status = RtlAnsiStringToUnicodeString(ImportNameUnic, &ImportNameAnsi, FALSE);
+                if (NT_SUCCESS(Status))
+                {
+                    RedirectedDll = FALSE;
+                    RedirectedImportName = ImportNameUnic;
+
+                    /* Check if the SxS Assemblies specify another file */
+                    Status = RtlDosApplyFileIsolationRedirection_Ustr(TRUE,
+                                                                      ImportNameUnic,
+                                                                      &LdrApiDefaultExtension,
+                                                                      UpdateString,
+                                                                      NULL,
+                                                                      &RedirectedImportName,
+                                                                      NULL,
+                                                                      NULL,
+                                                                      NULL);
+                    /* Check success */
+                    if (NT_SUCCESS(Status))
+                    {
+                        if (ShowSnaps)
+                        {
+                            DPRINT1("LDR: %Z got redirected to %wZ\n", &ImportNameAnsi, RedirectedImportName);
+                        }
+                        /* Let Ldrp know */
+                        RedirectedDll = TRUE;
+                    }
+
+                    if (RedirectedDll || Status == STATUS_SXS_KEY_NOT_FOUND)
```

Comment:
This whole block of logic is repeated at least twice, could be refactored in to a helper?

Reply:
I refrained from doing any kind of refactoring as this code was already duplicated.